fedora_messaging/cli.py

```diff
@@ -155,3 +164,120 @@ def consume(exchange, queue_name, routing_key, callback, app_name):
+            api.publish(msg, publish_exchange)
+        except exceptions.PublishReturned as e:
+            _log.error("Unable to publish message: %s", str(e.reason))
+            sys.exit(1)
+        except exceptions.ConnectionException as e:
+            _log.error("Unable to connect to the queue: %s", str(e.reason))
+            sys.exit(1)
+
+
+@cli.command()
+@click.option("--file", help=_record_file_help, required=True, type=click.File("w"))
+@click.option("--limit", help=_limit_help, type=click.IntRange(1))
+@click.option("--app-name", help=_app_name_help)
+@click.option("--routing-key", help=_routing_key_help, multiple=True)
+@click.option("--queue-name", help=_queue_name_help)
+@click.option("--exchange", help=_exchange_help)
+def record(exchange, queue_name, routing_key, app_name, limit, file):
+    """Record messages from an AMQP queue to provided file."""
+
+    class Recorder:
+        def __init__(self):
+            self.counter = 0
+            self.messages = []
+
+            def _signal_handler(signum, frame):
+                """
+                Signal handler that gracefully terminates the recorder
+
+                Args:
+                    signum (int): The signal this process received.
+                    frame (frame): The current stack frame (unused).
+                """
+                if signum in (signal.SIGTERM, signal.SIGINT):
+                    self.save_messages_to_file()
+                    raise exceptions.HaltConsumer(exit_code=0, requeue=False)
+
+            signal.signal(signal.SIGTERM, _signal_handler)
+            signal.signal(signal.SIGINT, _signal_handler)
+
+        def __call__(self, message):
+            self.messages.append(message)
+            self.counter += 1
+            if limit and limit == self.counter:
+                self.save_messages_to_file()
+                raise exceptions.HaltConsumer(exit_code=0, requeue=False)
+
+        def save_messages_to_file(self):
+            try:
+                json_str = dumps(self.messages)
+            except TypeError as e:
+                _log.error("Unable to save messages to file: ", str(e))
+            else:
+                file.write(json_str)
+
+    if exchange and queue_name and routing_key:
+        bindings = [
+            {"exchange": exchange, "queue": queue_name, "routing_keys": routing_key}
+        ]
+    elif not exchange and not queue_name and not routing_key:
+        bindings = config.conf["bindings"]
+    else:
+        raise click.ClickException(
+            "You must define all three of exchange, queue_name and"
+            " routing_key, or none of them to use the configuration"
+        )
+    if not bindings:
+        raise click.ClickException(
+            "No bindings are defined in the configuration file"
+            " and none were provided as arguments!"
+        )
+
+    if app_name:
+        config.conf["client_properties"]["app"] = app_name
+
+    _log.info("Recording to file %s started.", file)
```

Comment:
As mentioned above, you probably want to use click.echo here (or the progress bar)